pkg/operator/controller/dnsnameresolver/controller.go

```diff
@@ -0,0 +1,217 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ipRemovalGracePeriod = time.Second * 5
+
+	controllerName = "dnsnameresolver_controller"
+)
+
+// NewUnmanaged creates and returns a controller that watches DNSNameResolver
+// objects. The controller re-resolves the DNS names which get added to the
+// status of the DNSNameResolver objects. It also removes IP addresses of DNS
+// names, whose TTLs have expired, from the staus. This is an unmanaged
+// controller, which means that the manager does not start it.
+func NewUnmanaged(mgr manager.Manager, config Config) (controller.Controller, []cache.Cache, error) {
+	// Create a new cache for tracking the DNSNameResolver resources in
+	// the DNSNameResolverNamespace.
+	dnsNameResolverCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+		DefaultNamespaces: map[string]cache.Config{
+			config.DNSNameResolverNamespace: {},
+		},
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create a new cache to track the EndpointSlices corresponding to the
+	// CoreDNS pods.
+	corednsEndpointsSliceCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+		DefaultNamespaces: map[string]cache.Config{
+			config.OperandNamespace: {},
+		},
+		DefaultLabelSelector: labels.SelectorFromSet(labels.Set{
+			discoveryv1.LabelServiceName: config.ServiceName,
+		}),
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Initialize the reconciler.
+	reconciler := &reconciler{
+		config:               config,
+		dnsNameResolverCache: dnsNameResolverCache,
+		client:               mgr.GetClient(),
+		res:                  NewResolver(corednsEndpointsSliceCache, config.ServiceName),
+	}
+
+	// Create an unmanaged controller using the reconciler.
+	c, err := controller.NewUnmanaged(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Watch for the DNSNameResolver resources using dnsNameResolverCache.
+	if err := c.Watch(source.Kind(dnsNameResolverCache, &networkv1alpha1.DNSNameResolver{}), &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, nil, err
+	}
+
+	// Watch for the CoreDNS pod EndpointSlices using corednsEndpointsSliceCache.
+	if err := c.Watch(source.Kind(corednsEndpointsSliceCache, &discoveryv1.EndpointSlice{}), handler.EnqueueRequestsFromMapFunc(skipReconcileForEndpointSlice)); err != nil {
+		return nil, nil, err
+	}
+
+	return c, []cache.Cache{dnsNameResolverCache, corednsEndpointsSliceCache}, nil
+}
+
+// skipReconcileForEndpointSlice just skips the reconcile if any event related to the CoreDNS pod
+// endpointslices is received.
+func skipReconcileForEndpointSlice(context.Context, client.Object) []reconcile.Request {
+	return nil
+}
```

Comment:
```suggestion
	// Watch for the CoreDNS pod EndpointSlices to keep corednsEndpointsSliceCache synced. No reconcile requests should be generated.
	if err := c.Watch(source.Kind(corednsEndpointsSliceCache, &discoveryv1.EndpointSlice{}), handler.EnqueueRequestsFromMapFunc(handler.MapFunc{ return nil })); err != nil {
		return nil, nil, err
	}

	return c, []cache.Cache{dnsNameResolverCache, corednsEndpointsSliceCache}, nil
}
```
I don't think you need `skipReconcileForEndpointSlice`. In my opinion, an anonymous function is better for readability here.

I think it would also help to have a comment explicitly stating that the reason for this watch is only to keep `corednsEndpointsSliceCache` updated. Maybe it's a common pattern elsewhere, but when I first read this, I thought it was an error that no reconcile requests were generated.

Reply:
> Maybe it's a common pattern elsewhere, but when I first read this, I thought it was an error that no reconcile requests were generated.

I am actually not sure if it's used elsewhere. There were 2 options that I could think of for retrieving the CoreDNS pod IPs when sending the DNS lookup requests. Either use a client or a cache.

IMO, the CoreDNS pod IPs are quite stable apart from the cases where nodes are added/removed. Using the client everytime seemed like an overkill for me.

For using the cache, there should be a watch on the CoreDNS pod EndpointSlices, but as these EndpointSlices are not related to the DNSNameResolver objects, triggering a reconcile is not required.

I am open to suggestions to get the CoreDNS pod IPs in any other way, if possible.

Reply:
Moved to openshift/coredns-ocp-dnsnameresolver#2